### PR TITLE
Integrate new slots into lzy

### PR DESCRIPTION
**Pull request by ArtoLord**

### CODE REVIEW

File: lzy/lzy-service/src/main/java/ai/lzy/service/operations/start/StartExecution.java [hunk 1]
```diff
@@ -37,12 +29,7 @@ private StartExecution(StartExecutionBuilder builder) {
         super(builder);
         this.storageConfig = builder.storageConfig;
         this.state = builder.state;
-        this.allocatorSessionSpec = builder.allocatorSessionSpec;
-        this.portalServiceSpec = builder.portalServiceSpec;
-        this.abClient = builder.abClient;
-        this.allocOpClient = builder.allocOpClient;
-        this.steps = List.of(createKafkaTopic(), createAllocatorSession(), createPortalSubject(),
-            startAllocationPortalVm(), waitAllocationPortalVm(), StartExecution.this::complete);
+        this.steps = List.of(createKafkaTopic(), StartExecution.this::complete);
```
Comment (imakunin) [inline lzy/lzy-service/src/main/java/ai/lzy/service/operations/start/StartExecution.java:32]: maybe we should create allocator session here and pass it to the graph executor

File: lzy/slots/src/main/java/ai/lzy/fs/Slots.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+public class Slots {
+    private final Path fsRoot;
+    private final SlotsService slotsService;
+    private final List<SlotsExecutionContext> contexts = new ArrayList<>();  // Guarded by this
+    private final Supplier<String> tokenSupplier;
```
Comment (imakunin) [inline lzy/slots/src/main/java/ai/lzy/fs/Slots.java:29]: just `token`

File: lzy/worker/src/test/java/ai/lzy/worker/WorkerTests.java [hunk 3]
```diff
@@ -1,382 +0,0 @@
...
-        var slotsApiStub = LzySlotsApiGrpc.newBlockingStub(slotsApiChannel);
-        var slotsApiOpStub = LongRunningServiceGrpc.newBlockingStub(slotsApiChannel);
-
-        return new WorkerDesc(workerId, worker, workerKeys, workerJwt,
-            workerApiChannel, workerApiStub, workerApiOpStub,
-            slotsApiChannel, slotsApiStub, slotsApiOpStub);
-    }
-
-    @Test
-    public void apiAvailability() throws Exception {
```
Comment (imakunin) [inline lzy/worker/src/test/java/ai/lzy/worker/WorkerTests.java:285]: don't remove these tests, plz

File: lzy/slots/src/main/java/ai/lzy/fs/Slots.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            serviceName, iamChannel);
+
+        server = GrpcUtils.newGrpcServer(slotsApiAddress, authInterceptor)
+            .addService(ServerInterceptors.intercept(slotsService, interceptor))
+            .build();
+
+        server.start();
+    }
+
+    public synchronized SlotsExecutionContext context(
```
Comment (imakunin) [inline lzy/slots/src/main/java/ai/lzy/fs/Slots.java:67]: codestyle

**Review by imakunin — CHANGES_REQUESTED** [review]

**Review by zhvkgj — CHANGES_REQUESTED** [review]
you should delete portal specific files from `ai.lzy.service.config` package

**Review by zhvkgj — CHANGES_REQUESTED** [review]
`ai.lzy.service.dao.DataFlowGraph` should not have portal references and `Data's` suppliers and consumers are not slot names but slot ids

**Review by zhvkgj — CHANGES_REQUESTED** [review]

File: lzy/lzy-service/src/main/java/ai/lzy/service/dao/impl/ExecutionDaoImpl.java [hunk 5]
```diff
@@ -83,10 +82,7 @@ INSERT INTO workflow_executions (execution_id, user_id, created_at, storage, sto
     private static final String QUERY_SELECT_STOP_EXECUTION_DATA = """
         SELECT
             kafka_topic_json,
-            allocator_session_id,
-            portal_subject_id,
-            portal_vm_id,
-            portal_vm_address
+            allocator_session_id
```
[the remark this answers is not in the document]
Comment (ArtoLord) [reply lzy/lzy-service/src/main/java/ai/lzy/service/dao/impl/ExecutionDaoImpl.java:57]: Step is returned, so keeping it

File: iam-api/src/main/java/ai/lzy/iam/grpc/interceptors/AllowSubjectOnlyInterceptor.java [hunk 6]
```diff
@@ -53,25 +53,47 @@ public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(ServerCall<ReqT, Re
...
+        String clientName, Channel iamChannel)
+    {
+        var client = new AccessServiceGrpcClient(clientName, iamChannel, () -> new JwtCredentials("stub"));
+
+        return new AllowSubjectOnlyInterceptor(auth -> permissions.entrySet().stream().allMatch(entry -> {
+            var resource = entry.getKey();
+            var permission = entry.getValue();
+
+            try {
+                return client
```
[the remark this answers is not in the document]
Comment (ArtoLord) [reply iam-api/src/main/java/ai/lzy/iam/grpc/interceptors/AllowSubjectOnlyInterceptor.java:88]: We do not have batch handle in iam on this request.
I think we can add it, but in other task.

File: lzy/lzy-service/src/main/java/ai/lzy/service/BeanFactory.java
Comment (zhvkgj) [inline lzy/lzy-service/src/main/java/ai/lzy/service/BeanFactory.java:1]: the service still has need a iam's client beans and long-running channel manager?

File: lzy/lzy-service/src/main/java/ai/lzy/service/dao/DataFlowGraph.java [hunk 7]
```diff
@@ -243,7 +243,7 @@ private List<Data> calculateDataFlow() {
                 var supplierSlotName = pair.getValue().getKey();
                 var consumerSlotNamesAndOpIds = dataConsumers.get(supplierSlotUri);
 
-                List<String> consumerSlotNames = null;
+                List<String> consumerSlotNames = new ArrayList<>();
 
                 if (consumerSlotNamesAndOpIds != null) {
                     consumerSlotNames = new ArrayList<>(consumerSlotNamesAndOpIds.size());
```
Comment (zhvkgj) [inline lzy/lzy-service/src/main/java/ai/lzy/service/dao/DataFlowGraph.java:249]: its redundant if you've already initialized variable

**Review by zhvkgj — CHANGES_REQUESTED** [review]

File: iam-api/src/main/java/ai/lzy/iam/grpc/interceptors/AllowSubjectOnlyInterceptor.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (imakunin) [reply iam-api/src/main/java/ai/lzy/iam/grpc/interceptors/AllowSubjectOnlyInterceptor.java:88]: make a task, plz

File: lzy/lzy-service/src/main/java/ai/lzy/service/operations/start/StartExecution.java [hunk 8]
```diff
@@ -37,12 +31,8 @@ private StartExecution(StartExecutionBuilder builder) {
         super(builder);
         this.storageConfig = builder.storageConfig;
         this.state = builder.state;
+        this.steps = List.of(createKafkaTopic(), StartExecution.this::complete);
```
Comment (imakunin) [inline lzy/lzy-service/src/main/java/ai/lzy/service/operations/start/StartExecution.java:34]: `createAllocatorSession()` step is missing

File: lzy/lzy-service/src/main/java/ai/lzy/service/operations/stop/AbortExecution.java [hunk 9]
```diff
@@ -15,8 +15,7 @@ public final class AbortExecution extends StopExecution {
     public AbortExecution(AbortExecutionBuilder builder) {
         super(builder);
         this.graphClient = builder.graphClient;
-        this.steps = List.of(stopGraphs(), finishPortal(), waitFinishPortal(), freePortalVm(), deleteAllocSession(),
-            waitDeleteAllocSession(), deletePortalSubject(), destroyChannels(), deleteKafkaTopic(), this::complete);
+        this.steps = List.of(stopGraphs(), destroyChannels(), deleteKafkaTopic(), this::complete);
```
Comment (imakunin) [inline lzy/lzy-service/src/main/java/ai/lzy/service/operations/stop/AbortExecution.java:18]: `deleteAllocSession()` step is missing

File: lzy/lzy-service/src/main/java/ai/lzy/service/operations/stop/FinishExecution.java [hunk 10]
```diff
@@ -1,38 +1,26 @@
...
 public final class FinishExecution extends StopExecution {
-    private final LongRunningServiceBlockingStub channelsOpService;
     private final List<Supplier<StepResult>> steps;
 
     private FinishExecution(FinishExecutionBuilder builder) {
         super(builder);
-        this.channelsOpService = builder.channelsOpClient;
-        this.steps = List.of(finishPortal(), waitFinishPortal(), freePortalVm(), deleteAllocSession(),
-            waitDeleteAllocSession(), deletePortalSubject(), destroyChannels(), deleteKafkaTopic(), this::complete);
+        this.steps = List.of(destroyChannels(), deleteKafkaTopic(), this::complete);
```
Comment (imakunin) [inline lzy/lzy-service/src/main/java/ai/lzy/service/operations/stop/FinishExecution.java:14]: `deleteAllocSession()` step is missing

File: lzy/worker/src/test/java/ai/lzy/worker/WorkerTests.java [hunk 11]
```diff
@@ -210,12 +210,6 @@ LzySlotsApiGrpc.LzySlotsApiBlockingStub slotsApiStub(@Nullable String token) {
             return newBlockingClient(slotsApiStub, "x", () -> creds);
         }
 
-        LongRunningServiceGrpc.LongRunningServiceBlockingStub slotsApiOpStub(@Nullable String token) {
```
Comment (imakunin) [inline lzy/worker/src/test/java/ai/lzy/worker/WorkerTests.java:213]: do we no longer have an OperationService for slots?

File: lzy/worker/src/test/java/ai/lzy/worker/WorkerTests.java [hunk 12]
```diff
@@ -336,43 +320,33 @@ public void apiAvailability() throws Exception {
...
-            Assert.assertEquals(e.toString(), Status.Code.PERMISSION_DENIED, e.getStatus().getCode());
-
-            // SlotsApi Ops available for internal user
+            // Worker Ops available for internal user
             e = assertThrows(StatusRuntimeException.class, () ->
                 worker.workerApiOpStub(internalUser).get(GetOperationRequest.newBuilder().setOperationId("1").build()));
             Assert.assertEquals(e.toString(), Status.Code.NOT_FOUND, e.getStatus().getCode());
 
-            // SlotsApi unavailable for worker
+            // SlotsApi available for worker in same workflow
```
Comment (imakunin) [inline lzy/worker/src/test/java/ai/lzy/worker/WorkerTests.java:328]: why?

File: lzy/lzy-service/src/main/java/ai/lzy/service/operations/start/StartExecution.java (hunk 8 above)
Comment (ArtoLord) [reply lzy/lzy-service/src/main/java/ai/lzy/service/operations/start/StartExecution.java:34]: I didn't add this step here to not create fictive sessions for now.

**Review by imakunin — APPROVED** [review]

**Review by zhvkgj — APPROVED** [review]

**Pull request merged**